### ISSUE TITLE
Improvement regarding attributes syntax

### ISSUE THREAD

**Issue by john-cj** [asciidoctor/asciidoctor#3069]
I would suggest two possible improvements regarding attributes syntax: the first is about commas, and the second is about naming.

Consider this table:

```
[options="autowidth, header" role="role1 role2"]
|===
| Operation      | Input values | Output value
| Addition       | 5, 5         | 10
| Multiplication | 5, 5         | 25
|===

++++
<style>
.role1 { opacity: .5; }
.role2 { font-style: italic; }
</style>
++++
```

There is a required comma between `autowidth` and `header`, there could be an optional comma before `role`, and there must be no comma between `role1` and `role2`. It's not very easy to remember. Isn't it better to make all of the commas optional or drop all of them?

Then, naming:

* Most attibutes have full names, and sometimes there is an additional shortened form: `options`, `opts`, `role`, `grid`. However, `cols` doesn't have full name. Isn't it better to add it?
* `options` and `cols` are plural because it's possible to have multiple ones. Isn't it better to add `roles` and use it as default in manual? (Singular `role` could be used as an alias for backward compatibility).

**Comment by mojavelinux** [comment]
The reason role is different is because it maps to CSS class in HTML and role in DocBook (both singular). And both those attributes take a space-separated list, so no conversion is necessary. All other attributes that take multiple values use a comma.

Btw, you can shorten how you write options as follows:

```
[%header%footer]
|===
|===
```

> However, `cols` doesn't have full name. Isn't it better to add it?

The idea is to keep attribute names roughly the same length. Again, this is a convention that comes from both HTML and DocBook.

**Comment by john-cj** [comment]
Yes, thanks about mentioned shortened syntax (`%option.role`). Sometimes I use it, very handy.

> The reason role is different is because it maps to CSS class in HTML and role in DocBook (both singular). And both those attributes take a space-separated list, so no conversion is necessary. All other attributes that take multiple values use a comma.

As I understand, the advantage of current approach is in slightly higher speed of conversion. However, this technical background is unknown to a user, and therefore current approach looks just inconsistently. Also, it's easier to introduce comma-related errors. So, a bit of performance could be sacrificed to the convenience of end users (just my opinion).

I like Asciidoctor with patience, that's why I'm dogged in some critique.

> The idea is to keep attribute names roughly the same length. Again, this is a convention that comes from both HTML and DocBook.

But we already have `options` and `opts`. It doesn't look any different from `columns` and `cols`.

**Comment by mojavelinux** [comment]
The reason we have opts is because options is inconsistent (in length). That's why we shortened it in Asciidoctor. `cols` is very well established in markup languages, from XML (DITA, DocBook) to HTML. It's `columns` that would be inconsistent, IMO.

**Comment by john-cj** [comment]
There are different types of consistency (aka accuracy):

* to use the same length (`cols`, `grid`, `opts`)
* to use the established conventions from DocBook (singular `role`) and CSS (shortened `cols`)
* to use the same common sense logic (plural `roles` as a twin of `options`, unshortened `columns` as a twin of `options`, and allow commas between roles since they are allowed/required everywhere else).

And of course, there are different types of people with different personal aesthetic views. The good side is that the third type of consistency could be achieved without breaking anything in the previous two, with the aliases.

**Comment by mojavelinux** [comment]
One of the central goals of AsciiDoc is to build on existing conventions. That's what's gotten us to where we are and that's a strategy we're going to continue using. The goal is not to have puristic consistency just for consistency sake, as that would actually work against existing mental models. In other words, it's not dogmatic. It's practical.

What I'm might be open to is allowing commas between role values. And the reason I say that is because we already do that for inline elements. These would be optional commas that would get replaced with spaces during parsing. I don't think that would be unreasonable.

I'm not willing to change attribute names at this point. It adds far too much complexity both to the parsing and the perception of the language.

**Comment by elextr** [comment]
Clearly its not a suitable change for the 2.0.0 release, but its important to consider that many Asciidoc users do not have knowledge of CSS, or Docbook, they use Asciidoc to avoid that.  

So being consistent with those output formats is not important to them.  And as Asciidoc grows its user base, that cohort will grow.

For them being consistent with plain English is something that needs to be considered in the future.

**Comment by john-cj** [comment]
Just for the record, another comma-related issue that might be confusing:

Example 1: works

```
[%autowidth%header.stretch, cols="3", width=50%]
|===
| aaa
| b
| c

| aaa
| b
| c

| aaa
| b
| c
|===
```

Example 2: Works

```
[%autowidth%header.stretch, cols="3" width=50%]
```

Example 3: Doesn't work

```
[%autowidth%header.stretch cols="3" width=50%]
```
